# Programming
This is our first Jupyter Notebook. Notebooks are a way to interactively explore computing concepts without installing a lot of software. Basically, they act like a website where you can type in some code and run it. In this particular notebook, we will explore some very basic programming concepts. You'll learn enough Python to treat it like a fancy calculator. Hopefully, you'll also get the hang of Jupyter Notebooks. Let's get started.

## Literals
Literal represent values meant to be interpreted exactly as they're written. There's nothing profound or fancy going on here. For example, 2 represents the number two. If we input this number, Python outputs that number. While this may not be very exciting, we have to start somewhere.

The box below this text is a Jupyter Notebook cell. It's set up to be an interactive Python session. You can type some Python code into it and run then it. To type code in, just click anywhere in the cell and then start typing. When you're done, you can click the "Run" button above to see the output. Try typing in the literal 2 and running the program.

In [ ]:
2

Hopefully, that worked. If it did, your input of 2 should produce an output 2 below it. Feel free to experiment within Jupyter cells. Type in something different and run that instead. Playing around is the best way to learn. If you encounter any difficulties along the way, reach out to me, and I'll do what I can to get you back on track.

## Expressions
Literals are data, and data doesn't do anything on its own. If we could only use literals in our computer programs, computers would not be very useful. Fortunately, we have ways to manipulate data. **Operators** are symbols like + and - that allow us to interact with data and produce new values. By combining literals and operators, we can create **expressions**. Expressions are combinations of symbols that we can evaluate to produce a value. Expressions look a bit like simple math equations.

There are two expressions in the cells below. If you run the two programs (click the cell and then click the "Run" button above,) I don't expect the resulting output values will surprise you. Again, don't be afraid to change the code in the cells and see what happens. It's the best way to learn.

In [ ]:
2+8

In [ ]:
4-2

## Variables and assignment
Even if we combine literals and operators into expressions, we can't produce very powerful programs. The programs we've seen so far look like things a simple calculator could manage. We  gain a little more flexibility and power by storing the results of our expressions in **variables**. Variables allow us to give a name to a value and store it in memory. We assign a value to a variable using a special operator called the **assignment operator**, which looks like an equals sign =. Below we have an assignment statement. Running this cell produces no output, because we stored the value in memory.

In [ ]:
a=6+2

Note that the variable name *always* goes on the left side of the equals sign, and the expression *always* goes on the right side. We stored the value 8 in memory. We can now use it.

In [ ]:
a-4

Let's come back to that line we wrote above: a=6+2. It's important. That was our first **statement**. A statement is a line of code that changes the state of the computer. After encountering the line a=6+2, the computer must *remember* what the variable stores. A computer **program** is a series of statements. Each statement sets up the computer to process subsequent statements.

Here's an example program to figure out how many years it took for people to go from designing the first computer to being able to buy one as a consumer product.

In [ ]:
yearAnalyticEngine=1833
yearAtari8080=1975
yearAtari8080-yearAnalyticEngine

Note that the value of the last expression we type is printed whenever we run our program. This is a feature of Jupyter notebooks. There are other ways to produce output, but we'll worry about those a little bit later. Let's focus more on operators for now.

## Some other operators

There are many operators in Python that probably work the way you would expect.

|Operator | Name                | Description                     | &nbsp; &nbsp; &nbsp; Equation       |
| ------- | ------------------- | ------------------------------- | -------------- |
| x + y   | Addition            | Adds x and y                    | $x+y$          |
| x - y   | Subtraction         | Subtracts y from x              | $x-y$          |
| x * y   | Multiplication      | Multiplies x and y              | $x\times y$    |
| x / y   | Division            | Divides x by y.                 | $\frac{x}{y}$  |
| x\*\*y  | Exponentiation      | Multiplies x by itself y times  | $x^y$          |

We can combine them into complicated expressions like the one below, which calculates $y=3x^2+3x+4$ when $x=15$. Oh no! It's a quadratic equation! High school algebra is back! Run! Don't be too scared. I'll make sure to review any math you need in this class.

In [ ]:
x=15
y=3*x**2+3*x+4
y

Note that we *always* have to include the operators in Python. $3x$ might be shorthand for $3 \times x$ in math class, but in Python, it's just gibberish. If you run the code below, you'll see an error message. "invalid syntax" is how the Python interpreter says "I have no idea what you're trying to say."

In [ ]:
3x


## Order of Operations

Wait a minute. How do we know it's $3x^2$ and not $(3x)^2$? How do we know to multiply after exponentiation? Just like in math class, all the operators we'll learn have an order of precedence. Good news! **I promise to never require you to learn them.** If it's ever unclear, just ask me.

Just like in Math class, we can put parentheses around the operations we want to come first. This overrides the order of operations. $(3+2)*4$ is not the same as $3+(2*4)$ in math. They're not the same in Python either.

In [ ]:
(3+2)*4

In [ ]:
3+(2*4)

If we just use parentheses, we can always be clear about what order of operations we intend. The example below might be ugly and a bit overkill, but it certainly makes the order of operations in our quadratic equation clear. Remember: start with the innermost parentheses and work your way out.

In [ ]:
y=(((3*(x**2))+(3*x))+4)
y

## Functions

*Functions* (sometimes called *subprograms* or *subroutines*) are pre-written code that we can use in our program. Later, we will learn to write our own functions, but for now we will take advantage of some built-in Python functions that come premade. To run or **invoke** a function, we use its name followed by matching parentheses. Between the parentheses, we can put the input to the function (called **arguments**.)

As an example, the **print** function prints the input to the screen.

In [2]:
print(1)
print(2)

1
2


Now, we can produce output anytime we like. We can provide multiple arguments to a function by separating them with parenthesis.

In [3]:
print(1,2,3)

1 2 3


There are other built-in functions, like **min** and **max** that find the min and max of their arguments. They **return** a result, meaning invoking these functions produces a value, just like an expression.

In [7]:
x=min(5,10,20)
y=max(10,3,7)
print(x,y)

5 10


## Modulo

There are other operators in Python that you'll probably find less intuitive. Let's explore one of them.

|Operator | Name                | Description                         | &nbsp; &nbsp; &nbsp; Equation   
| ------- | ------------------- | ----------------------------------- | -------------------- |
| x // y  | Integer division    | Divide x by y, but ignore the decimal part | $ \lfloor\frac{x}{y}\rfloor $|
| x % y   | Modulo              | The remainder of x divided by y    | $ x\mod y $|

Remember remainders? You might have to think back to primary school. It might seem strange, but modulo is arguably one of the most important operators we'll learn. Make sure you have a good intuition for how it works. There's an example in the cell below using $\frac{23}{8}$:

In [22]:
a=23/8
b=23//8
c=23%8
print(a,b,c)

2.875 2 7


If we did long division of $\frac{23}{8}$ by hand or with a calculator, we'd get 2.875. If we instead we think back to the tricks we learned about fractions, we'd remember $\frac{23}{8}=\frac{8}{8}+\frac{8}{8}+\frac{7}{8}=2+\frac{7}{8}$, or 2 remainder 7.

## Exercises
Okay. Here we go. Your first programming exercises. Do your best to work through these on your own. If you find yourself puzzling over any question for more than 10 minutes, take a break and come back to it later. If you still can't figure it out, ask course staff for help.

1) Write an expression that sums all the prime numbers between 1 and 20.

2) Write an expression to compute the area of a circle with radius 4.2. Remember that the equation for the area of a circle is $\pi r^2$ and $\pi \approx 3.14159$.

In [ ]:
r=4.2
pi=3.14159

3) Write an expression to figure out if the number 7,810,124 is divisible by 9.

4) If I start waiting at exactly noon for 2,295,331 minutes, what number will the minute hand point to? Write a Python expression to figure this out.

5) It's Sunday. What day will it be in 38,292 days? Write a Python expression to figure this out.

6) Compute the value of ${(\frac{16}{7}})^{9}+(4-\frac{3}{2})$ and store the result in a variable named *myAnswer*.

7) Write an expression to find the largest remainder we get when dividing the numbers from 126 to 130 by 13.